core/src/test/java/com/graphhopper/routing/weighting/custom/SpeedCalculatorTest.java

```diff
@@ -0,0 +1,184 @@
+import com.graphhopper.routing.profiles.EnumEncodedValue;
+import com.graphhopper.routing.profiles.RoadClass;
+import com.graphhopper.routing.profiles.RoadEnvironment;
+import com.graphhopper.routing.util.CustomModel;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.storage.GraphBuilder;
+import com.graphhopper.util.EdgeIteratorState;
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class SpeedCalculatorTest {
+
+    private final EncodingManager em;
+    private final EdgeIteratorState edge;
+
+    SpeedCalculatorTest() {
+        // in this test we use the same edge for every test and check the calculated speed for different custom models
+        // and depending on additional properties of the edge we set in the tests
+        em = EncodingManager.create("car");
+        edge = new GraphBuilder(em).create().edge(0, 1, 1000, true);
+    }
+
+    @Test
+    public void maxSpeedFallback() {
+        // default speed for car is 60km/h
+        assertEquals(60, calcSpeed(edge, new CustomModel()));
+        // use global fallback to limit speed
+        assertEquals(30, calcSpeed(edge, new CustomModel().setMaxSpeedFallback(30.0)));
+    }
+
+    @Test
+    public void maxSpeed() {
+        // here we use max_speed to limit speed
+        CustomModel model = new CustomModel();
+        Map<String, Object> map = new HashMap<>();
+        map.put("*", 25);
+        model.getMaxSpeed().put(RoadClass.KEY, map);
+        assertEquals(25, calcSpeed(edge, model));
+        // ... but setting it to a higher value than the default has no effect
+        map.put("*", 70);
+        assertEquals(60, calcSpeed(edge, model));
+        // also our edge might or might not be affected
+        // todonow: why has encoding manager a road class encoded value by default?
+        EnumEncodedValue<RoadClass> roadClass = em.getEnumEncodedValue(RoadClass.KEY, RoadClass.class);
+        edge.set(roadClass, RoadClass.PRIMARY);
+        // todonow: why do we not allow an empty map as value?
+        model.getMaxSpeed().clear();
+        assertEquals(60, calcSpeed(edge, new CustomModel()));
+        map.clear();
+        map.put(RoadClass.PRIMARY.toString(), 10);
+        model.getMaxSpeed().put(RoadClass.KEY, map);
+        assertEquals(10, calcSpeed(edge, model));
+    }
+
+    @Test
+    public void invalidMaxSpeed() {
+        CustomModel model = new CustomModel();
+        Map<String, Object> map = new HashMap<>();
+        map.put("*", 300);
+        model.getMaxSpeed().put(RoadClass.KEY, map);
+        try {
+            assertEquals(60, calcSpeed(edge, model));
+            fail();
+        } catch (IllegalArgumentException e) {
+            // todonow: why do we not allow this, setting max speed to 300 simply has no effect and the user cannot really
+            // know that the max speed is 140
```

Comment:
@karussell ?

Reply:
The problem is that the user might think that he can increase the max speed this way. But of course it is ugly that the user would have to know the maximum speed for every vehicle. 
Should we silently reduce this to the value of the vehicle?

Reply:
Ok lets leave it as it is for now :)